src/apps/eth/eth_sign_msg.c

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "eth_sign_msg.h"
+#include "eth_common.h"
+#include "eth_params.h"
+#include "eth_sighash.h"
+#include "eth_verify.h"
+#include <sha3.h>
+
+#include <hardfault.h>
+#include <keystore.h>
+#include <util.h>
+#include <workflow/confirm.h>
+
+app_eth_sign_error_t app_eth_sign_msg(const ETHSignMessageRequest* request, ETHSignResponse* response)
+{
+    const app_eth_coin_params_t* params = app_eth_params_get(request->coin);
+    if (params == NULL) {
+        return APP_ETH_SIGN_ERR_INVALID_INPUT;
+    }
+    if (!eth_common_is_valid_keypath(request->coin, request->keypath, request->keypath_count)) {
+        return APP_ETH_SIGN_ERR_INVALID_INPUT;
+    }
+    if(request->msg.size > 1024) {
+        Abort("Invalid message size");
+    }
+    const char msg_header[] = "\x19" "Ethereum Signed Message:\n";
+    // sizeof(msg_header) includes null terminator
+    // the maximum length of the signed data is 1024, therefore 4 bytes might be needed as length
+    // prefix.
+    char msg[sizeof(msg_header) + sizeof(request->msg.bytes) + 4] = {0};
+    strcat(msg, msg_header);
```

Comment:
wouldn't it be possible to add the header as an arg in the snprintf below? then the strcat and null terminator care might be dropped.

Reply:
done